### PR TITLE
No error on toplevel return in JS

### LINKED ISSUE #48224: TS1108  "problem" showing on pure Javascript (node) file - Erroneous Error Flag
Issue Type: <b>Bug</b>

PROBLEMS pane shows:
   [JS] file.js (1)
      A 'return' statement can only be used within a function body.  ts(1108) [400,5]

My node Javascript file is statements that call defined functions.  These statements have an if statement that executes a return statement when true.  The program is Javascript, and recognized as such by editor.   There is no typescript in the environment. This bug appeared immediately after VS Code updated to February 2022 version (1.65.2).  It did not appear in the previous version with the exact same file.  The program runs properly on the target when run.

UPDATE: I have installed 1.64.2 to remediate issue.  The problem does not occur down-level.

The program looks like this:

#!/usr/bin/env node
statements and function definitions
if ( x == y)
  return          
statements

A Typescript error should not show for a Javascript file.

VS Code version: Code 1.65.2 (c722ca6c7eed3d7987c0d5c3df5c45f6b15e77d1, 2022-03-10T14:33:55.248Z)
OS version: Windows_NT x64 10.0.19043
Restricted Mode: No

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i7-6770HQ CPU @ 2.60GHz (8 x 2592)|
|GPU Status|2d_canvas: enabled<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>oop_rasterization: enabled<br>opengl: enabled_on<br>rasterization: enabled<br>skia_renderer: enabled_on<br>video_decode: enabled<br>vulkan: disabled_off<br>webgl: enabled<br>webgl2: enabled|
|Load (avg)|undefined|
|Memory (System)|31.89GB (23.34GB free)|
|Process Argv||
|Screen Reader|no|
|VM|0%|
</details><details><summary>Extensions (11)</summary>

Extension|Author (truncated)|Version
---|---|---
Bookmarks|ale|13.2.4
bracket-pair-colorizer|Coe|1.0.62
html-preview-vscode|geo|0.2.5
vscode-pull-request-github|Git|0.38.1
ibm-jcl|kel|0.11.0
remote-wsl|ms-|0.64.2
rexx|nex|0.0.1
deepdark-material|Nim|3.3.0
ayu|tea|1.0.5
pdf|tom|1.2.0
vscode-extension-for-zowe|Zow|1.22.0

(17 theme extensions excluded)

</details><details>
<summary>A/B Experiments</summary>

```
vsliv368cf:30146710
vsreu685:30147344
python383cf:30185419
vspor879:30202332
vspor708:30202333
vspor363:30204092
pythonvspyl392:30443607
pythontb:30283811
pythonvspyt551cf:30345471
pythonptprofiler:30281270
vshan820:30294714
vstes263:30335439
vscorecescf:30445987
pythondataviewer:30285071
vscod805:30301674
pythonvspyt200:30340761
binariesv615:30325510
bridge0708:30335490
bridge0723:30353136
vsaa593:30376534
vsc1dst:30438360
pythonvs932:30410667
wslgetstarted:30449410
vsclayoutctrt:30451275
dsvsc009:30440023
pythonvspyt640:30450904
vscscmwlcmc:30438804
vscgsvid2:30447481
pynewfile477:30450038

```

</details>

### PR DESCRIPTION
**Pull request by sandersn**
Turns out it's only an error in modules. It's possible to keep this error on the list of "OK for JS" errors and make the checker code stop issuing it for JS scripts only. However, I don't think the error is valuable enough to do that.

Fixes #48224

### CODE REVIEW

File: tests/cases/conformance/salsa/plainJSGrammarErrors.ts [hunk 1]
```diff
@@ -208,5 +208,3 @@ function foo() { new.targe }
 const nullaryDynamicImport = import()
 const trinaryDynamicImport = import('1', '2', '3')
 const spreadDynamicImport = import(...[])
-
-return
```
Comment (jakebailey) [inline tests/cases/conformance/salsa/plainJSGrammarErrors.ts:212]: Should we be leaving this return here so that we don't accidentally readd this error?

**Comment by gabritto** [comment]
Is detecting if a js file is a script or a module not reliable? I'm not sure I understand why you feel it is not valuable enough to do this

**Comment by sandersn** [comment]
Making decisions based on JS-vs-TS is not a common thing to do in the middle of the checker. Of course, there are some constructs like commonjs imports that *should* only show up in JS, but the checker treats them the same regardless of location. In this case, it would have to skip the error only in the case that top-level `return` is in a JS file.

**Comment by sandersn** [comment]
Alternatively, we could instead stop issuing the error for any script. It's really good default assumption that a `.ts` script will emit to `.js` script that will run in node or the browser, where top-level return is legal.

Actually, I kind of like that idea. Let me try it.

**Comment by gabritto** [comment]
> Alternatively, we could instead stop issuing the error for any script. It's really good default assumption that a `.ts` script will emit to `.js` script that will run in node or the browser, where top-level return is legal.
> 
> Actually, I kind of like that idea. Let me try it.

yeah, that's what I was thinking and it seems reasonable to me

File: tests/baselines/reference/multiLinePropertyAccessAndArrowFunctionIndent1.types [hunk 2]
```diff
@@ -6,12 +6,12 @@ return this.edit(role)
 >this.edit : any
 >this : typeof globalThis
 >edit : any
->role : any
+>role : error
```
Comment (jakebailey) [inline tests/baselines/reference/multiLinePropertyAccessAndArrowFunctionIndent1.types:9]: This is a weird change... this is the `role` parameter in this test's confusing return statement. Not sure why it went from an implicit any to an `error`.

File: tests/baselines/reference/showConfig/Shows tsconfig for single option/noDtsResolution/tsconfig.json [hunk 3]
```diff
@@ -1,5 +0,0 @@
-{
-    "compilerOptions": {
```
Comment (DanielRosenwasser) [inline tests/baselines/reference/showConfig/Shows tsconfig for single option/noDtsResolution/tsconfig.json:2]: Why did this get deleted?

File: tests/baselines/reference/parserStatementIsNotAMemberVariableDeclaration1.errors.txt [hunk 4]
```diff
@@ -1,16 +0,0 @@
...
-!!! error TS1108: A 'return' statement can only be used within a function body.
-    
-      "set": function (key, value) {
-    
-        // 'private' should not be considered a member variable here.
-        private[key] = value;
-    
-      }
-    
-    };
```
Comment (DanielRosenwasser) [inline tests/baselines/reference/parserStatementIsNotAMemberVariableDeclaration1.errors.txt:16]: We're still supposed to disallow this in TS files, right?

**Review by DanielRosenwasser — CHANGES_REQUESTED** [review]

**Comment by DanielRosenwasser** [comment]
Allowing `return` at the top-level of a `.ts` file was never allowed, and probably shouldn't be allowed now. Node.js is one of the few environments that actually allows this, and new code should use `process.exit(0)` to work across environments.

**Comment by sandersn** [comment]
Hmm, OK. I looked again and there are piles of `isInJSFile` calls throughout the checker, so I guess I made up the "don't check for js" rule there.

I'll push a commit that makes the checker issue the error for all non-js files and for JS files that use ES modules.

File: src/compiler/checker.ts [hunk 5]
```diff
@@ -38913,7 +38913,9 @@ namespace ts {
             }
 
             if (!container) {
-                grammarErrorOnFirstToken(node, Diagnostics.A_return_statement_can_only_be_used_within_a_function_body);
+                if (!!getSourceFileOfNode(node).externalModuleIndicator || !isInJSFile(node)) {
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts:38916]:
```suggestion
                if (!isInJSFile(node)) {
```

Comment (DanielRosenwasser) [reply src/compiler/checker.ts:38916]: In fact, I would say that the right fix is still to issue the error in checkJs, but not when providing the default grammar errors for JS files.

**Review by DanielRosenwasser — CHANGES_REQUESTED** [review]
Just undo the introduction of this check. It used to only show up in TS files and in JS files when checkJS was on. Reinstate that behavior.

File: tests/baselines/reference/showConfig/Shows tsconfig for single option/noDtsResolution/tsconfig.json (hunk 3 above)
Comment (sandersn) [reply tests/baselines/reference/showConfig/Shows tsconfig for single option/noDtsResolution/tsconfig.json:2]: I'm pretty sure this is an orphaned baseline. I have no idea how I decided to delete it though.

**Review by jakebailey — APPROVED** [review]
This LGTM now, a conservative choice, but I don't know if @DanielRosenwasser or @gabritto have any other last comments. I know Danial's PR status is still requesting changes.

**Comment by sandersn** [comment]
I'm pretty sure I addressed Daniel's comments; Daniel, Gabriela and I will need to decide what future semantics should be separately, so I'm going to merge this PR so that it's not hanging around until the last minute of the beta.

**Pull request merged**